### PR TITLE
Proper threadsafe collection of stream snapshots

### PR DESCRIPTION
**Pull request by johanandren**
Also handles stopping streams better (ignoring them rather than timing out and retrying)

Refs #28960

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala [hunk 1]
```diff
@@ -229,34 +225,78 @@ private[akka] class SubFusingActorMaterializerImpl(
...
+  timers.startSingleTimer(Timeout, Timeout, timeout)
+
+  override def receive: Receive = {
+    case snap: StreamSnapshot =>
+      collected = snap :: collected
+      leftToRespond -= sender()
+      completeIfDone()
+    case Terminated(streamActor) =>
+      leftToRespond -= streamActor
+      completeIfDone()
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala:289]: If a stream stopped between supervisor extracting the list and getting a snapshot, we just ignore that child (I think that makes most sense)

File: akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala [hunk 2]
```diff
@@ -229,34 +225,78 @@ private[akka] class SubFusingActorMaterializerImpl(
...
+    case SnapshotTimeout =>
+      replyTo ! StatusReply.Error(
+        s"Didn't get replies from all stream actors within the timeout of ${timeout.toMillis} ms")
+      context.stop(self)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala:293]: Possible improvement would be to report what we got so far if this happens. Didn't want to invest that much in this fix though.

File: akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala [hunk 3]
```diff
@@ -229,34 +225,78 @@ private[akka] class SubFusingActorMaterializerImpl(
   implicit val ec: ExecutionContextExecutor = context.dispatcher
   override def supervisorStrategy: SupervisorStrategy = SupervisorStrategy.stoppingStrategy
 
+  private var snapshotCollectors: Set[ActorRef] = Set.empty
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala:228]: We need a way to filter out the collector children from the streams (without adding extra overhead to stream materialization)

File: akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala [hunk 4]
```diff
@@ -229,34 +225,78 @@ private[akka] class SubFusingActorMaterializerImpl(
...
     case GetChildren =>
       sender() ! Children(context.children.toSet)
-    case GetChildrenSnapshots =>
-      takeSnapshotsOfChildren().map(ChildrenSnapshots.apply _).pipeTo(sender())
-
+    case GetChildrenSnapshots(timeout) =>
+      val collector =
+        context.actorOf(
+          SnapshotCollector
+            .props(context.children.toSet -- snapshotCollectors, timeout, sender())
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala:240]: `diff` is supposed to be better than `--` for Sets

Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala:240]: If other actors are named we could get away without the `snapshotCollectors` collection by filtering out children with name starting with `$`. Maybe too sneaky.

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala (hunk 4 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala:240]: I think they all are anonymous, and I didn't want to introduce a counter to keep collector children names unique.

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala:240]: Oh wait, I'm wrong, the name comes from the materialize command. Hmm. Maybe this is more clear anyway.

File: akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala [hunk 5]
```diff
@@ -229,34 +225,78 @@ private[akka] class SubFusingActorMaterializerImpl(
...
+    streamActors.foreach { ref =>
+      context.watch(ref)
+      ref ! ActorGraphInterpreter.Snapshot
     }
+  }
 
-    // If the timeout hits it is likely because one of the streams stopped between looking at the list
-    // of children and asking it for a snapshot. We retry the entire snapshot in that case
-    retry(() => takeSnapshot(), 3)
+  timers.startSingleTimer(Timeout, Timeout, timeout)
```
Comment (ennru) [inline akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala:280]: Shouldn't this send `SnapshotTimeout`?

**Review by ennru — APPROVED** [review]
LGTM, just one glitch I believe.

File: akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala (hunk 5 above)
Comment (octonato) [reply akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala:280]: Indeed, this should be `ShapshotTimeout`

Just commenting here to increase visibility. PR is approved already. We don't want to miss this.

**Review by octonato — APPROVED** [review]

File: akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala (hunk 2 above)
Comment (octonato) [reply akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala:293]: What will it take to report it? 

Is it relevant to mention the number of stream actors that didn't reported on time? If that's only this, why not adding it?

File: akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala (hunk 5 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala:280]: Pesky classic actors...

**Pull request merged**